File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/extensions/Models/ModelTypes.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public types for model management.
 * These are thin wrappers over C++ types in rac_model_types.h
 * Business logic (format support, capability checks) is in C++.
 *
 * Mirrors Swift ModelTypes.swift exactly.
 */

package com.runanywhere.sdk.public.extensions.Models

import com.runanywhere.sdk.public.extensions.LLM.ThinkingTagPattern
import kotlinx.serialization.Serializable

// MARK: - Model Source

/**
 * Source of model data (where the model info came from).
 * Mirrors Swift ModelSource exactly.
 */
@Serializable
enum class ModelSource(
    val value: String,
) {
    /** Model info came from remote API (backend model catalog) */
    REMOTE("remote"),

    /** Model info was provided locally via SDK input (addModel calls) */
    LOCAL("local"),
}

// MARK: - Model Format

/**
 * Model formats supported.
 * Mirrors Swift ModelFormat exactly.
 */
@Serializable
enum class ModelFormat(
    val value: String,
) {
    ONNX("onnx"),
    ORT("ort"),
    GGUF("gguf"),
    BIN("bin"),
    UNKNOWN("unknown"),
}

// MARK: - Model Selection Context

/**
 * Context for model selection UI - determines which models to show.
 * Mirrors Swift ModelSelectionContext exactly.
 */
@Serializable
enum class ModelSelectionContext(
    val value: String,
) {
    /** Select a language model (LLM) */
    LLM("llm"),

    /** Select a speech-to-text model */
    STT("stt"),

    /** Select a text-to-speech model/voice */
    TTS("tts"),

    /** Select models for voice agent (all 3 types) */
    VOICE("voice"),

    /** Select an embedding model for RAG (ONNX) */
    RAG_EMBEDDING("ragEmbedding"),

    /** Select an LLM for RAG generation (llama.cpp) */
    RAG_LLM("ragLLM"),

    /** Select a vision language model (VLM) */
    VLM("vlm"),
    ;

    /** Human-readable title for the selection context */
    val title: String
        get() =
            when (this) {
                LLM -> "Select LLM Model"
                STT -> "Select STT Model"
                TTS -> "Select TTS Voice"
                VOICE -> "Select Voice Models"
                RAG_EMBEDDING -> "Select Embedding Model"
                RAG_LLM -> "Select LLM Model"
                VLM -> "Select Vision Model"
            }

    /** Check if a category is relevant for this selection context */
    fun isCategoryRelevant(category: ModelCategory): Boolean =
        when (this) {
            LLM -> category == ModelCategory.LANGUAGE
            STT -> category == ModelCategory.SPEECH_RECOGNITION
            TTS -> category == ModelCategory.SPEECH_SYNTHESIS
            VOICE ->
                category == ModelCategory.LANGUAGE ||
                    category == ModelCategory.SPEECH_RECOGNITION ||
                    category == ModelCategory.SPEECH_SYNTHESIS
            RAG_EMBEDDING -> category == ModelCategory.EMBEDDING
            RAG_LLM -> category == ModelCategory.LANGUAGE
            VLM ->
                category == ModelCategory.MULTIMODAL ||
                    category == ModelCategory.VISION
        }

    /** Check if a framework is relevant for this selection context */
    fun isFrameworkRelevant(framework: com.runanywhere.sdk.core.types.InferenceFramework): Boolean =
        when (this) {
            LLM ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.LLAMA_CPP ||
                    framework == com.runanywhere.sdk.core.types.InferenceFramework.FOUNDATION_MODELS
            STT ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.ONNX
            TTS ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.ONNX ||
                    framework == com.runanywhere.sdk.core.types.InferenceFramework.SYSTEM_TTS ||
                    framework == com.runanywhere.sdk.core.types.InferenceFramework.FLUID_AUDIO
            VOICE ->
                LLM.isFrameworkRelevant(framework) ||
                    STT.isFrameworkRelevant(framework) ||
                    TTS.isFrameworkRelevant(framework)
            RAG_EMBEDDING ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.ONNX
            RAG_LLM ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.LLAMA_CPP
            VLM ->
                framework == com.runanywhere.sdk.core.types.InferenceFramework.LLAMA_CPP
        }
}

// MARK: - Model Category

/**
 * Defines the category/type of a model based on its input/output modality.
 * Mirrors Swift ModelCategory exactly.
 */
@Serializable
enum class ModelCategory(
    val value: String,
) {
    LANGUAGE("language"), // Text-to-text models (LLMs)
    SPEECH_RECOGNITION("speech-recognition"), // Voice-to-text models (ASR)
    SPEECH_SYNTHESIS("speech-synthesis"), // Text-to-voice models (TTS)
    VISION("vision"), // Image understanding models
    IMAGE_GENERATION("image-generation"), // Text-to-image models
    MULTIMODAL("multimodal"), // Models that handle multiple modalities
    AUDIO("audio"), // Audio processing (diarization, etc.)
    EMBEDDING("embedding"), // Embedding models (RAG, semantic search)
    ;

    /** Whether this category typically requires context length */
    val requiresContextLength: Boolean
        get() = this == LANGUAGE || this == MULTIMODAL

    /** Whether this category typically supports thinking/reasoning */
    val supportsThinking: Boolean
        get() = this == LANGUAGE || this == MULTIMODAL
}

// MARK: - Archive Types

/**
 * Supported archive formats for model packaging.
 * Mirrors Swift ArchiveType exactly.
 */
@Serializable
enum class ArchiveType(
    val value: String,
) {
    ZIP("zip"),
    TAR_BZ2("tar.bz2"),
    TAR_GZ("tar.gz"),
    TAR_XZ("tar.xz"),
    ;

    /** File extension for this archive type */
    val fileExtension: String get() = value

    companion object {
        /** Detect archive type from URL path */
        fun from(path: String): ArchiveType? {
            val lowercased = path.lowercase()
            return when {
                lowercased.endsWith(".tar.bz2") || lowercased.endsWith(".tbz2") -> TAR_BZ2
                lowercased.endsWith(".tar.gz") || lowercased.endsWith(".tgz") -> TAR_GZ
                lowercased.endsWith(".tar.xz") || lowercased.endsWith(".txz") -> TAR_XZ
                lowercased.endsWith(".zip") -> ZIP
                else -> null
            }
        }
    }
}

/**
 * Describes the internal structure of an archive after extraction.
 * Mirrors Swift ArchiveStructure exactly.
 */
@Serializable
enum class ArchiveStructure(
    val value: String,
) {
    SINGLE_FILE_NESTED("singleFileNested"),
    DIRECTORY_BASED("directoryBased"),
    NESTED_DIRECTORY("nestedDirectory"),
    UNKNOWN("unknown"),
}

// MARK: - Expected Model Files

/**
 * Describes what files are expected after model extraction/download.
 * Mirrors Swift ExpectedModelFiles exactly.
 */
@Serializable
data class ExpectedModelFiles(
    val requiredPatterns: List<String> = emptyList(),
    val optionalPatterns: List<String> = emptyList(),
    val description: String? = null,
) {
    companion object {
        val NONE = ExpectedModelFiles()
    }
}

/**
 * Describes a file that needs to be downloaded as part of a multi-file model.
 * Mirrors Swift ModelFileDescriptor exactly.
 */
@Serializable
data class ModelFileDescriptor(
    /** Full URL to download this file from */
    val url: String,
    /** Filename to save as (e.g., "model.gguf" or "mmproj.gguf") */
    val filename: String,
    /** Whether this file is required for the model to work */
    val isRequired: Boolean = true,
) {
    /** Legacy compatibility */
    val relativePath: String get() = url.substringAfterLast('/').substringBefore('?')
    val destinationPath: String get() = filename
}

// MARK: - Model Artifact Type

/**
 * Describes how a model is packaged and what processing is needed after download.
 * Mirrors Swift ModelArtifactType exactly.
 */
@Serializable
sealed class ModelArtifactType {
    @Serializable
    data class SingleFile(
        val expectedFiles: ExpectedModelFiles = ExpectedModelFiles.NONE,
    ) : ModelArtifactType()

    @Serializable
    data class Archive(
        val archiveType: ArchiveType,
        val structure: ArchiveStructure,
        val expectedFiles: ExpectedModelFiles = ExpectedModelFiles.NONE,
    ) : ModelArtifactType()

    @Serializable
    data class MultiFile(
        val files: List<ModelFileDescriptor>,
    ) : ModelArtifactType()

    @Serializable
    data class Custom(
        val strategyId: String,
    ) : ModelArtifactType()

    @Serializable
    data object BuiltIn : ModelArtifactType()

    val requiresExtraction: Boolean
        get() = this is Archive

    val requiresDownload: Boolean
        get() = this !is BuiltIn

    val expectedFilesValue: ExpectedModelFiles
        get() =
            when (this) {
                is SingleFile -> expectedFiles
                is Archive -> expectedFiles
                else -> ExpectedModelFiles.NONE
            }

    val displayName: String
        get() =
            when (this) {
                is SingleFile -> "Single File"
                is Archive -> "${archiveType.value.uppercase()} Archive"
                is MultiFile -> "Multi-File (${files.size} files)"
                is Custom -> "Custom ($strategyId)"
                is BuiltIn -> "Built-in"
            }

    companion object {
        /** Infer artifact type from download URL */
        @Suppress("UNUSED_PARAMETER")
        fun infer(url: String?, format: ModelFormat): ModelArtifactType {
            // format parameter reserved for future use when format-specific inference is needed
            if (url == null) return SingleFile()
            val archiveType = ArchiveType.from(url)
            return if (archiveType != null) {
                Archive(archiveType, ArchiveStructure.UNKNOWN)
            } else {
                SingleFile()
            }
        }
    }
}

// MARK: - Model Info

/**
 * Information about a model - in-memory entity.
 * Mirrors Swift ModelInfo exactly.
 */
@Serializable
data class ModelInfo(
    // Essential identifiers
    val id: String,
    val name: String,
    val category: ModelCategory,
    // Format and location
    val format: ModelFormat,
    val downloadURL: String? = null,
    var localPath: String? = null,
    // Artifact type
    val artifactType: ModelArtifactType = ModelArtifactType.SingleFile(),
    // Size information
    val downloadSize: Long? = null,
    // Framework
    val framework: com.runanywhere.sdk.core.types.InferenceFramework,
    // Model-specific capabilities
    val contextLength: Int? = null,
    val supportsThinking: Boolean = false,
    val supportsLora: Boolean = false,
    val thinkingPattern: ThinkingTagPattern? = null,
    // Optional metadata
    val description: String? = null,
    // Tracking fields
    val source: ModelSource = ModelSource.REMOTE,
    val createdAt: Long = System.currentTimeMillis(),
    var updatedAt: Long = System.currentTimeMillis(),
) {
    /** Whether this model is downloaded and available locally */
    val isDownloaded: Boolean
        get() {
            val path = localPath ?: return false
            if (path.startsWith("builtin://")) return true
            // Actual file check would be done in platform-specific code
            return path.isNotEmpty()
        }

    /** Whether this model is available for use */
    val isAvailable: Boolean get() = isDownloaded

    /** Whether this is a built-in platform model */
    val isBuiltIn: Boolean
        get() {
            if (artifactType is ModelArtifactType.BuiltIn) return true
            val path = localPath
            if (path != null && path.startsWith("builtin://")) return true
            return framework == com.runanywhere.sdk.core.types.InferenceFramework.FOUNDATION_MODELS ||
                framework == com.runanywhere.sdk.core.types.InferenceFramework.SYSTEM_TTS
        }
}

// MARK: - Download Progress

/**
 * Progress information for model downloads.
 */
@Serializable
data class DownloadProgress(
    /** Model ID being downloaded */
    val modelId: String,
    /** Progress percentage (0.0 to 1.0) */
    val progress: Float,
    /** Bytes downloaded so far */
    val bytesDownloaded: Long,
    /** Total bytes to download (null if unknown) */
    val totalBytes: Long?,
    /** Download state */
    val state: DownloadState,
    /** Error message if state is ERROR */
    val error: String? = null,
)

/**
 * State of a model download.
 */
@Serializable
enum class DownloadState {
    PENDING,
    DOWNLOADING,
    EXTRACTING,
    COMPLETED,
    ERROR,
    CANCELLED,
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Models/ModelTypes.swift
//
//  ModelTypes.swift
//  RunAnywhere SDK
//
//  Public types for model management.
//  These are thin wrappers over C++ types in rac_model_types.h
//  Business logic (format support, capability checks) is in C++.
//

import CRACommons
import Foundation

// MARK: - Model Source

/// Source of model data (where the model info came from)
public enum ModelSource: String, Codable, Sendable {
    /// Model info came from remote API (backend model catalog)
    case remote

    /// Model info was provided locally via SDK input (addModel calls)
    case local
}

// MARK: - Model Format

/// Model formats supported
public enum ModelFormat: String, CaseIterable, Codable, Sendable {
    case onnx
    case ort
    case gguf
    case bin
    case coreml
    case unknown
}

// MARK: - Model Category

/// Defines the category/type of a model based on its input/output modality
public enum ModelCategory: String, CaseIterable, Codable, Sendable {
    case language = "language"              // Text-to-text models (LLMs)
    case speechRecognition = "speech-recognition"  // Voice-to-text models (ASR)
    case speechSynthesis = "speech-synthesis"      // Text-to-voice models (TTS)
    case vision = "vision"                  // Image understanding models
    case imageGeneration = "image-generation"      // Text-to-image models
    case multimodal = "multimodal"          // Models that handle multiple modalities
    case audio = "audio"                    // Audio processing (diarization, etc.)
    case embedding = "embedding"            // Embedding models (RAG, semantic search)

    /// Whether this category typically requires context length
    /// Note: C++ equivalent is rac_model_category_requires_context_length()
    public var requiresContextLength: Bool {
        switch self {
        case .language, .multimodal:
            return true
        case .speechRecognition, .speechSynthesis, .vision, .imageGeneration, .audio, .embedding:
            return false
        }
    }

    /// Whether this category typically supports thinking/reasoning
    /// Note: C++ equivalent is rac_model_category_supports_thinking()
    public var supportsThinking: Bool {
        switch self {
        case .language, .multimodal:
            return true
        case .speechRecognition, .speechSynthesis, .vision, .imageGeneration, .audio, .embedding:
            return false
        }
    }
}

// MARK: - Inference Framework

/// Supported inference frameworks/runtimes for executing models
public enum InferenceFramework: String, CaseIterable, Codable, Sendable {
    // Model-based frameworks
    case onnx = "ONNX"
    case llamaCpp = "LlamaCpp"
    case foundationModels = "FoundationModels"
    case systemTTS = "SystemTTS"
    case fluidAudio = "FluidAudio"
    case coreml = "CoreML"        // Core ML (Apple Neural Engine) for diffusion models
    case mlx = "MLX"              // MLX (Apple Silicon VLM via MLX C++)
    case whisperKitCoreML = "WhisperKitCoreML" // WhisperKit CoreML (Apple Neural Engine) for STT

    // Special cases
    case builtIn = "BuiltIn"      // For simple services (e.g., energy-based VAD)
    case none = "None"            // For services that don't use a model
    case unknown = "Unknown"      // For unknown/unspecified frameworks

    /// Human-readable display name for the framework
    public var displayName: String {
        switch self {
        case .onnx: return "ONNX Runtime"
        case .llamaCpp: return "llama.cpp"
        case .foundationModels: return "Foundation Models"
        case .systemTTS: return "System TTS"
        case .fluidAudio: return "FluidAudio"
        case .coreml: return "Core ML"
        case .mlx: return "MLX"
        case .whisperKitCoreML: return "WhisperKit CoreML"
        case .builtIn: return "Built-in"
        case .none: return "None"
        case .unknown: return "Unknown"
        }
    }

    /// Snake_case key for analytics/telemetry
    public var analyticsKey: String {
        switch self {
        case .onnx: return "onnx"
        case .llamaCpp: return "llama_cpp"
        case .foundationModels: return "foundation_models"
        case .systemTTS: return "system_tts"
        case .fluidAudio: return "fluid_audio"
        case .coreml: return "coreml"
        case .mlx: return "mlx"
        case .whisperKitCoreML: return "whisperkit_coreml"
        case .builtIn: return "built_in"
        case .none: return "none"
        case .unknown: return "unknown"
        }
    }
}

// MARK: - InferenceFramework C++ Bridge

public extension InferenceFramework {
    /// Convert Swift InferenceFramework to C rac_inference_framework_t
    func toCFramework() -> rac_inference_framework_t {
        switch self {
        case .onnx: return RAC_FRAMEWORK_ONNX
        case .llamaCpp: return RAC_FRAMEWORK_LLAMACPP
        case .foundationModels: return RAC_FRAMEWORK_FOUNDATION_MODELS
        case .systemTTS: return RAC_FRAMEWORK_SYSTEM_TTS
        case .fluidAudio: return RAC_FRAMEWORK_FLUID_AUDIO
        case .coreml: return RAC_FRAMEWORK_COREML
        case .mlx: return RAC_FRAMEWORK_MLX
        case .whisperKitCoreML: return RAC_FRAMEWORK_WHISPERKIT_COREML
        case .builtIn: return RAC_FRAMEWORK_BUILTIN
        case .none: return RAC_FRAMEWORK_NONE
        case .unknown: return RAC_FRAMEWORK_UNKNOWN
        }
    }

    /// Create Swift InferenceFramework from C rac_inference_framework_t
    static func fromCFramework(_ cFramework: rac_inference_framework_t) -> InferenceFramework {
        switch cFramework {
        case RAC_FRAMEWORK_ONNX: return .onnx
        case RAC_FRAMEWORK_LLAMACPP: return .llamaCpp
        case RAC_FRAMEWORK_FOUNDATION_MODELS: return .foundationModels
        case RAC_FRAMEWORK_SYSTEM_TTS: return .systemTTS
        case RAC_FRAMEWORK_FLUID_AUDIO: return .fluidAudio
        case RAC_FRAMEWORK_COREML: return .coreml
        case RAC_FRAMEWORK_MLX: return .mlx
        case RAC_FRAMEWORK_WHISPERKIT_COREML: return .whisperKitCoreML
        case RAC_FRAMEWORK_BUILTIN: return .builtIn
        case RAC_FRAMEWORK_NONE: return .none
        default: return .unknown
        }
    }

    /// Initialize from a string, matching case-insensitively.
    init?(caseInsensitive string: String) {
        let lowercased = string.lowercased()

        if let exact = InferenceFramework(rawValue: string) {
            self = exact
            return
        }

        if let framework = InferenceFramework.allCases.first(where: { $0.rawValue.lowercased() == lowercased }) {
            self = framework
            return
        }

        if let framework = InferenceFramework.allCases.first(where: { $0.analyticsKey == lowercased }) {
            self = framework
            return
        }

        return nil
    }
}

// MARK: - Archive Types

/// Supported archive formats for model packaging
public enum ArchiveType: String, CaseIterable, Codable, Sendable {
    case zip = "zip"
    case tarBz2 = "tar.bz2"
    case tarGz = "tar.gz"
    case tarXz = "tar.xz"

    /// File extension for this archive type
    public var fileExtension: String {
        rawValue
    }

    /// Detect archive type from URL
    /// Note: C++ equivalent is rac_archive_type_from_path()
    public static func from(url: URL) -> ArchiveType? {
        let path = url.path.lowercased()
        if path.hasSuffix(".tar.bz2") || path.hasSuffix(".tbz2") {
            return .tarBz2
        } else if path.hasSuffix(".tar.gz") || path.hasSuffix(".tgz") {
            return .tarGz
        } else if path.hasSuffix(".tar.xz") || path.hasSuffix(".txz") {
            return .tarXz
        } else if path.hasSuffix(".zip") {
            return .zip
        }
        return nil
    }
}

/// Describes the internal structure of an archive after extraction
public enum ArchiveStructure: String, Codable, Sendable, Equatable {
    case singleFileNested
    case directoryBased
    case nestedDirectory
    case unknown
}

// MARK: - Expected Model Files

/// Describes what files are expected after model extraction/download
public struct ExpectedModelFiles: Codable, Sendable, Equatable {
    public let requiredPatterns: [String]
    public let optionalPatterns: [String]
    public let description: String?

    public init(
        requiredPatterns: [String] = [],
        optionalPatterns: [String] = [],
        description: String? = nil
    ) {
        self.requiredPatterns = requiredPatterns
        self.optionalPatterns = optionalPatterns
        self.description = description
    }

    public static let none = ExpectedModelFiles()
}

/// Describes a file that needs to be downloaded as part of a multi-file model
public struct ModelFileDescriptor: Codable, Sendable, Equatable {
    /// Full URL to download this file from
    public let url: URL
    /// Filename to save as (e.g., "model.gguf" or "mmproj.gguf")
    public let filename: String
    /// Whether this file is required for the model to work
    public let isRequired: Bool

    public init(url: URL, filename: String, isRequired: Bool = true) {
        self.url = url
        self.filename = filename
        self.isRequired = isRequired
    }

    // Legacy compatibility
    public var relativePath: String { url.lastPathComponent }
    public var destinationPath: String { filename }
}

// MARK: - Model Artifact Type

/// Describes how a model is packaged and what processing is needed after download.
public enum ModelArtifactType: Codable, Sendable, Equatable {
    case singleFile(expectedFiles: ExpectedModelFiles = .none)
    case archive(ArchiveType, structure: ArchiveStructure, expectedFiles: ExpectedModelFiles = .none)
    case multiFile([ModelFileDescriptor])
    case custom(strategyId: String)
    case builtIn

    public var requiresExtraction: Bool {
        if case .archive = self { return true }
        return false
    }

    public var requiresDownload: Bool {
        if case .builtIn = self { return false }
        return true
    }

    public var expectedFiles: ExpectedModelFiles {
        switch self {
        case .singleFile(let expected), .archive(_, _, let expected):
            return expected
        default:
            return .none
        }
    }

    public var displayName: String {
        switch self {
        case .singleFile:
            return "Single File"
        case .archive(let type, _, _):
            return "\(type.rawValue.uppercased()) Archive"
        case .multiFile(let files):
            return "Multi-File (\(files.count) files)"
        case .custom(let strategyId):
            return "Custom (\(strategyId))"
        case .builtIn:
            return "Built-in"
        }
    }

    /// Infer artifact type from download URL
    /// Note: C++ equivalent is rac_artifact_infer_from_url()
    public static func infer(from url: URL?, format _: ModelFormat) -> ModelArtifactType {
        guard let url = url else {
            return .singleFile(expectedFiles: .none)
        }
        if let archiveType = ArchiveType.from(url: url) {
            return .archive(archiveType, structure: .unknown, expectedFiles: .none)
        }
        return .singleFile(expectedFiles: .none)
    }
}

// MARK: - ModelArtifactType Codable

extension ModelArtifactType {
    private enum CodingKeys: String, CodingKey {
        case type, archiveType, structure, expectedFiles, files, strategyId
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)

        switch type {
        case "singleFile":
            let expected = try container.decodeIfPresent(ExpectedModelFiles.self, forKey: .expectedFiles) ?? .none
            self = .singleFile(expectedFiles: expected)
        case "archive":
            let archiveType = try container.decode(ArchiveType.self, forKey: .archiveType)
            let structure = try container.decode(ArchiveStructure.self, forKey: .structure)
            let expected = try container.decodeIfPresent(ExpectedModelFiles.self, forKey: .expectedFiles) ?? .none
            self = .archive(archiveType, structure: structure, expectedFiles: expected)
        case "multiFile":
            let files = try container.decode([ModelFileDescriptor].self, forKey: .files)
            self = .multiFile(files)
        case "custom":
            let strategyId = try container.decode(String.self, forKey: .strategyId)
            self = .custom(strategyId: strategyId)
        case "builtIn":
            self = .builtIn
        default:
            self = .singleFile()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        switch self {
        case .singleFile(let expected):
            try container.encode("singleFile", forKey: .type)
            if expected != .none {
                try container.encode(expected, forKey: .expectedFiles)
            }
        case .archive(let archiveType, let structure, let expected):
            try container.encode("archive", forKey: .type)
            try container.encode(archiveType, forKey: .archiveType)
            try container.encode(structure, forKey: .structure)
            if expected != .none {
                try container.encode(expected, forKey: .expectedFiles)
            }
        case .multiFile(let files):
            try container.encode("multiFile", forKey: .type)
            try container.encode(files, forKey: .files)
        case .custom(let strategyId):
            try container.encode("custom", forKey: .type)
            try container.encode(strategyId, forKey: .strategyId)
        case .builtIn:
            try container.encode("builtIn", forKey: .type)
        }
    }
}

// MARK: - Model Info

/// Information about a model - in-memory entity
public struct ModelInfo: Codable, Sendable, Identifiable {
    // Essential identifiers
    public let id: String
    public let name: String
    public let category: ModelCategory

    // Format and location
    public let format: ModelFormat
    public let downloadURL: URL?
    public var localPath: URL?

    // Artifact type
    public let artifactType: ModelArtifactType

    // Size information
    public let downloadSize: Int64?

    // Framework
    public let framework: InferenceFramework

    // Model-specific capabilities
    public let contextLength: Int?
    public let supportsThinking: Bool
    public let thinkingPattern: ThinkingTagPattern?

    // Optional metadata
    public let description: String?

    // Tracking fields
    public let source: ModelSource
    public let createdAt: Date
    public var updatedAt: Date

    // MARK: - Computed Properties

    /// Whether this model is downloaded and available locally
    public var isDownloaded: Bool {
        guard let localPath = localPath else { return false }

        if localPath.scheme == "builtin" {
            return true
        }

        let (exists, isDirectory) = FileOperationsUtilities.existsWithType(at: localPath)

        if exists && isDirectory {
            return FileOperationsUtilities.isNonEmptyDirectory(at: localPath)
        }

        return exists
    }

    /// Whether this model is available for use
    public var isAvailable: Bool {
        isDownloaded
    }

    /// Whether this is a built-in platform model
    public var isBuiltIn: Bool {
        if artifactType == .builtIn {
            return true
        }
        if let localPath = localPath, localPath.scheme == "builtin" {
            return true
        }
        return framework == .foundationModels || framework == .systemTTS
    }

    private enum CodingKeys: String, CodingKey {
        case id, name, category, format, downloadURL, localPath
        case artifactType
        case downloadSize
        case framework
        case contextLength, supportsThinking, thinkingPattern
        case description
        case source, createdAt, updatedAt
    }

    public init(
        id: String,
        name: String,
        category: ModelCategory,
        format: ModelFormat,
        framework: InferenceFramework,
        downloadURL: URL? = nil,
        localPath: URL? = nil,
        artifactType: ModelArtifactType? = nil,
        downloadSize: Int64? = nil,
        contextLength: Int? = nil,
        supportsThinking: Bool = false,
        thinkingPattern: ThinkingTagPattern? = nil,
        description: String? = nil,
        source: ModelSource = .remote,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.format = format
        self.framework = framework
        self.downloadURL = downloadURL
        self.localPath = localPath

        self.artifactType = artifactType ?? ModelArtifactType.infer(from: downloadURL, format: format)

        self.downloadSize = downloadSize

        if category.requiresContextLength {
            self.contextLength = contextLength ?? 2048
        } else {
            self.contextLength = contextLength
        }

        self.supportsThinking = category.supportsThinking ? supportsThinking : false
        self.thinkingPattern = supportsThinking ? (thinkingPattern ?? .defaultPattern) : nil

        self.description = description
        self.source = source
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}
